internal/postgres/statement.go

```diff
@@ -0,0 +1,697 @@
+}
+
+func getTypeID(l []nodes.Node) (string, error) {
+	// The pg_query library generates a pg_catalog schema prefix for most
+	// types, but not for all. Typically "aliases" don't have the prefix.
+	// For example, "boolean" is parsed to ["pg_catalog", "bool"], but "bool" is
+	// parsed to ["bool"]. However the exact rules are unclear e.g. "date"
+	// is parsed to just ["date"].
+	// For simplicity, we strip off the pg_catalog prefix.
+	var ids []string
+	for _, t := range l {
+		s, err := getString(t)
+		if err != nil {
+			return "", err
+		}
+		ids = append(ids, s)
+	}
+	if len(ids) > 1 && ids[0] == "pg_catalog" {
+		ids = ids[1:]
+	}
+	return strings.Join(ids, "."), nil
+}
+
+func prType(pgCol pgColDef) string {
+	s := pgCol.id
+	if len(pgCol.mods) > 0 {
+		var l []string
+		for _, x := range pgCol.mods {
+			l = append(l, strconv.FormatInt(x, 10))
+		}
+		s = fmt.Sprintf("%s(%s)", s, strings.Join(l, ","))
+	}
+	if len(pgCol.array) > 0 {
+		l := []string{s}
+		for _, x := range pgCol.array {
+			if x == -1 {
+				l = append(l, "[]")
+			} else {
+				l = append(l, fmt.Sprintf("[%d]", x))
+			}
+		}
+		s = strings.Join(l, "")
+	}
+	return s
+}
+
+// getTableName extracts the table name from RangeVar n, and returns
+// the raw extracted name (the PostgreSQL table name) plus the corresponding
+// Spanner table name.
+func getTableName(conv *Conv, n nodes.RangeVar) (string, string, error) {
+	// RangeVar is used to represent table names. It consists of three components:
+	//  Catalogname: database name; either not specified or the current database
+	//  Schemaname: schemas are PostgreSql namepaces; often unspecified; defaults to "public"
+	//  Relname: name of the table
+	// We build a table name from these three components as follows:
+	// a) nil components are dropped.
+	// b) if more than one component is specified, they are joined using "."
+	//    (Note that Spanner doesn't allow "." in table names, so this
+	//    will eventually get re-mapped when we construct the Spanner table name).
+	// c) Schemaname is dropped if it is "public".
+	// d) return error if Relname is nil or "".
+	var l []string
+	if n.Catalogname != nil {
+		l = append(l, *n.Catalogname)
+	}
+	if n.Schemaname != nil && *n.Schemaname != "public" { // Don't include "public".
+		l = append(l, *n.Schemaname)
+	}
+	if n.Relname == nil && *n.Relname == "" {
+		return "", "", fmt.Errorf("relname is empty: can't build table name")
+	}
+	l = append(l, *n.Relname)
+	pgTable := strings.Join(l, ".")
+	spTable, err := GetSpannerTable(conv, pgTable)
+	if err != nil {
+		return "", "", err
+	}
+	return pgTable, spTable, nil
+}
+
+type constraint struct {
+	ct   nodes.ConstrType
+	keys []string
+}
+
+// analyzeConstraints traverses a list of nodes (expecting them to be
+// Constraint nodes), and collects the contraints they represent as
+// a list of constraint-type/key-list pairs.
+func analyzeConstraints(conv *Conv, n nodes.Node, pgTable string, l []nodes.Node) (cs []constraint) {
```

Comment:
Would extractConstraints make sense? It would make it clear what the return type is. To me, analyzeConstraints doesn't sound like it would return something.

Reply:
Done.
Yes, I think that is a more intuitive name.